dnf-behave-tests/features/repoquery/rich-deps.feature

```diff
@@ -0,0 +1,213 @@
+ Then the exit code is 0
+  And stdout is
+      """
+      c1-0:1.0-1.x86_64
+      """
+
+
+# a1-1.0: Conflicts: ((b1 and x1) or c1)
+Scenario: repoquery --whatconflicts for "((b1 and x1) or c1)"
+ When I execute dnf with args "repoquery --whatconflicts b1"
+ Then the exit code is 0
+  And stdout is
+      """
+      a1-0:1.0-1.x86_64
+      """
+ When I execute dnf with args "repoquery --whatconflicts c1"
+ Then the exit code is 0
+  And stdout is
+      """
+      a1-0:1.0-1.x86_64
+      """
+ When I execute dnf with args "repoquery --whatconflicts x1"
+ Then the exit code is 0
+  And stdout is
+      """
+      a1-0:1.0-1.x86_64
+      """
+Given I successfully execute dnf with args "install x1"
+ When I execute dnf with args "repoquery --whatconflicts b1"
+ Then the exit code is 0
+  And stdout is
+      """
+      a1-0:1.0-1.x86_64
+      """
+ When I execute dnf with args "repoquery --whatconflicts c1"
+ Then the exit code is 0
+  And stdout is
+      """
+      a1-0:1.0-1.x86_64
+      """
+ When I execute dnf with args "repoquery --whatconflicts x1"
+ Then the exit code is 0
+  And stdout is
+      """
+      a1-0:1.0-1.x86_64
+      """
+
+
+# a1-1.0: Recommends: (b1 < 2.0 if x1 >= 2.0 else c1)
+Scenario: repoquery --whatrecommends for "(b1 < 2.0 if x1 >= 2.0 else c1)"
+ When I execute dnf with args "repoquery --whatrecommends b1"
+ Then the exit code is 0
+  And stdout is
+      """
+      a1-0:1.0-1.x86_64
+      c1-0:1.0-1.x86_64
+      c1-0:2.0-1.x86_64
+      """
+ When I execute dnf with args "repoquery --whatrecommends 'b1 > 2.0'"
+ Then the exit code is 0
+  And stdout is
+      """
+      c1-0:1.0-1.x86_64
+      c1-0:2.0-1.x86_64
+      """
+ When I execute dnf with args "repoquery --whatrecommends x1"
+ Then the exit code is 0
+  And stdout is empty
+ When I execute dnf with args "repoquery --whatrecommends c1"
+ Then the exit code is 0
+  And stdout is
+      """
+      a1-0:1.0-1.x86_64
+      """
+
+
+# a1-1.0: Suggests: ((b1 with b1-prov2 > 1.7) or (c1 <= 1.0 without c1-prov1 > 0.5))
+Scenario: repoquery --whatsuggests for "((b1 with b1-prov2 > 1.7) or (c1 <= 1.0 without c1-prov1 > 0.5))"
+ When I execute dnf with args "repoquery --whatsuggests b1"
+ Then the exit code is 0
+  And stdout is
+      """
+      a1-0:1.0-1.x86_64
+      """
+ When I execute dnf with args "repoquery --whatsuggests c1"
+ Then the exit code is 0
+  And stdout is
+      """
+      a1-0:1.0-1.x86_64
+      """
+
+
+# a1-1.0: Supplements: ((b1 < 2.0 with b1-prov2 > 1.7) or (c1 > 1.0 without c1-prov1 > 0.5))
+# TODO neither of the above two expressions (the "with" and "without" ones) holds true,
+# but the two commands below still list a1 as if it matched
+# note: even if simplified to e.g. (b1 with nonexisting_provide) it still evaluates to true.
+@xfail
+Scenario: repoquery --whatsupplements for "((b1 < 2.0 with b1-prov2 > 1.7) or (c1 > 1.0 without c1-prov1 > 0.5))"
+ When I execute dnf with args "repoquery --whatsupplements b1"
```

Comment:
`repoquery --whatsupplements b1`
returns
`a1-0:1.0-1.x86_64`

Isn't it correct?

`repoquery --supplements a1-0:1.0-1.x86_64`
returns
`((b1 < 2.0 with b1-prov2 > 1.7) or (c1 > 1.0 without c1-prov1 > 0.5))`

The first condition is `true` because `--whatsupplements b1` was without a version.

For example `repoquery --whatsupplements 'b1 > 10'` returns nothing

Reply:
This may be a peculiarity of how the rich deps are resolved, but `b1 < 2.0` leaves only a single package, wich is `b1-1.0` and that one has a provide `b1-prov2 = 1.0`, which doesn't hold true for the second part of the `with` statement. So there is no `b1` package that fulfils the requirement.

But thinking about it now, this is probably how it works, since I'm not providing enough information in the query, so dnf lists what possibly supplements b1.

Reply:
Btw. this patch is related: https://github.com/openSUSE/libsolv/commit/85f82095c379ca46adecb6a137ce2bd47eb04c47

But AFAICS it won't handle this case because the dependencies aren't on the same provide name...